libcrux-ml-kem/src/serialize.rs

```diff
@@ -102,94 +131,156 @@ pub(super) fn deserialize_ring_elements_reduced<
         }
     }
     deserialized_pk
 }
 
 #[inline(always)]
+#[cfg_attr(hax, requires(
+    20 * (VECTORS_IN_RING_ELEMENT - 1) + 20 <= OUT_LEN
+))]
 fn compress_then_serialize_10<const OUT_LEN: usize, Vector: Operations>(
     re: &PolynomialRingElement<Vector>,
 ) -> [u8; OUT_LEN] {
     let mut serialized = [0u8; OUT_LEN];
     for i in 0..VECTORS_IN_RING_ELEMENT {
+        fstar!("assume (Libcrux_ml_kem.Vector.Traits.f_compress_pre #v_Vector
+            10l
+            (Libcrux_ml_kem.Vector.Traits.to_unsigned_representative #v_Vector
+                (re.Libcrux_ml_kem.Polynomial.f_coefficients.[ i ] <: v_Vector)
+            <:
+            v_Vector))");
         let coefficient =
             Vector::compress::<10>(to_unsigned_representative::<Vector>(re.coefficients[i]));
 
+        fstar!("assume (Libcrux_ml_kem.Vector.Traits.f_serialize_10_pre #v_Vector coefficient)");
         let bytes = Vector::serialize_10(coefficient);
         serialized[20 * i..20 * i + 20].copy_from_slice(&bytes);
     }
     serialized
 }
 
 #[inline(always)]
+#[cfg_attr(hax, requires(
+    22 * (VECTORS_IN_RING_ELEMENT - 1) + 22 <= OUT_LEN
+))]
 fn compress_then_serialize_11<const OUT_LEN: usize, Vector: Operations>(
     re: &PolynomialRingElement<Vector>,
 ) -> [u8; OUT_LEN] {
     let mut serialized = [0u8; OUT_LEN];
     for i in 0..VECTORS_IN_RING_ELEMENT {
+        fstar!("assume (Libcrux_ml_kem.Vector.Traits.f_compress_pre #v_Vector
+            11l
+            (Libcrux_ml_kem.Vector.Traits.to_unsigned_representative #v_Vector
+                (re.Libcrux_ml_kem.Polynomial.f_coefficients.[ i ] <: v_Vector)
+            <:
+            v_Vector))");
         let coefficient =
             Vector::compress::<11>(to_unsigned_representative::<Vector>(re.coefficients[i]));
 
+        fstar!("assume (Libcrux_ml_kem.Vector.Traits.f_serialize_11_pre #v_Vector coefficient)");
         let bytes = Vector::serialize_11(coefficient);
         serialized[22 * i..22 * i + 22].copy_from_slice(&bytes);
     }
     serialized
 }
 
 #[inline(always)]
+#[cfg_attr(hax, requires(
+    (COMPRESSION_FACTOR == 10 || COMPRESSION_FACTOR == 11) &&
+    (COEFFICIENTS_IN_RING_ELEMENT * COMPRESSION_FACTOR) / 8 == OUT_LEN
+))]
 pub(super) fn compress_then_serialize_ring_element_u<
     const COMPRESSION_FACTOR: usize,
     const OUT_LEN: usize,
     Vector: Operations,
 >(
     re: &PolynomialRingElement<Vector>,
 ) -> [u8; OUT_LEN] {
-    hax_debug_assert!((COEFFICIENTS_IN_RING_ELEMENT * COMPRESSION_FACTOR) / 8 == OUT_LEN);
-
     match COMPRESSION_FACTOR as u32 {
         10 => compress_then_serialize_10(re),
-        11 => compress_then_serialize_11(re),
-        _ => unreachable!(),
+        // Why it needs assumption here?
+        11 => {assume!(OUT_LEN == 352); compress_then_serialize_11(re)},
+        // unreachable!() does't verify, Option enum works but we have
+        // to edit refrences!!?
+        _ => [0u8; OUT_LEN],
```

Comment:
With Option enum path, we can the following
```pub(super) fn compress_then_serialize_ring_element_u<
    const COMPRESSION_FACTOR: usize,
    const OUT_LEN: usize,
    Vector: Operations,
>(
    re: &PolynomialRingElement<Vector>,
) -> Option<[u8; OUT_LEN]> {
    match COMPRESSION_FACTOR as u32 {
        10 => Some(compress_then_serialize_10(re)),
        11 => Some(compress_then_serialize_11(re)),
        _ => None,
    }
}```
This would work and verify but also need to modify the references!!

Reply:
This should stay `unreachable`. That's been the style used so far and worked well. Everything else is not correct.

Reply:
I am trying to verify `unreachable` avenue, if I couldn't I will seek help from Karthik.

Reply:
I replaced with `unreachable` expression and used proper asserts to complement that.